portalnet/src/discovery.rs

```diff
@@ -418,25 +481,10 @@ impl AsyncUdpSocket<UtpEnr> for Discv5UdpSocket {
     }
 
     async fn recv_from(&mut self, buf: &mut [u8]) -> io::Result<(usize, UtpEnr)> {
-        match self.talk_reqs.recv().await {
+        match self.talk_request_receiver.recv().await {
             Some(talk_req) => {
                 let src_node_id = talk_req.node_id();
-                let enr = match self.discv5.find_enr(src_node_id) {
-                    Some(enr) => UtpEnr(enr),
-                    None => {
-                        let enr = match self.discv5.cached_node_addr(src_node_id) {
-                            Some(node_addr) => Ok(node_addr.enr),
-                            None => {
-                                debug!(node_id = %src_node_id, "uTP packet from unknown source");
-                                Err(io::Error::new(
-                                    io::ErrorKind::Other,
-                                    "ENR not found for talk req destination",
-                                ))
-                            }
-                        }?;
-                        UtpEnr(enr)
-                    }
-                };
+                let enr = self.find_enr(src_node_id).await?;
```

Comment:
Can we determine the subprotocol making the request here, and pass that information into `find_enr`? Then `find_enr` can look more directly for the relevant ENR (if it needs the oracle), rather than hitting all of beacon, history, state, etc.

Reply:
There isn't any clear way to determine which Portal sub-protocol the uTP message is under, without a major refactor as currently all uTP connections are opaque. Other then all uTP connections currently being opaque on what Portal sub-network they are using.

One possible solution could be to share a Arc<RwLock<uTPEnrCache>> between all networks and this class the uTPEnrCache size would be determined by the max possible uTP connections that can be opened at a time. We put add the Enr whenever we would open a uTP connection. This way we wouldn't need to store which sub-network initiated the connection and would also not have to rely on calling the JSON-RPC. It would be weird that this Arc<> would basically be tangled everywhere though, but maybe that is fine.

If we think this solution is cleaner we can implement it in a follow up PR. Thinking more about it I do kind of like the solution I just suggested, but I think this would better be discussed on a meeting or something.

I am a little confused on what to do, but I do want to get something working out for the time being

Reply:
A issue with that approach though is we are possibly using more memory then we need to, so it is a jungling act of trade offs should we use less memory and worry less about being in-efficient or should we prioritize something in the middle I am not sure.

I think without proper performance testing (which wouldn't be the easiest to do or be a priority at this time) I am not sure if what I just proposed would even be worth it, because if we didn't do `utp_limit * 2 * active_networks` then we would still need to make JSON_RPC requests.

So on second thought I am not sure if it is that easy to come up with a best solution, only a group of solutions with trade offs